arangod/Transaction/CountCache.cpp

```diff
@@ -0,0 +1,53 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2016 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "CountCache.h"
+
+using namespace arangodb;
+using namespace arangodb::transaction;
+
+CountCache::CountCache() 
+   : count(CountCache::NotPopulated), 
+     timestamp(0.0) {}
+
+int64_t CountCache::get() const {
+  return count.load(std::memory_order_acquire);
+}
+
+int64_t CountCache::get(double ttl) const {
+  int64_t count = get();
+  if (count != CountCache::NotPopulated) {
+    double ts = timestamp.load(std::memory_order_acquire);
+    if (ts + Ttl > TRI_microtime()) {
+      // not yet expired
+      return count;
+    }
+  }
+  return CountCache::NotPopulated;
+}
+
+void CountCache::store(int64_t value) {
+  TRI_ASSERT(value >= 0);
+  timestamp.store(TRI_microtime(), std::memory_order_acquire);
```

Comment:
A store can never have acquire semantics (some implementations like the MSVC STL even have according assertions in debug builds).
The operations on `count` already have the correct memory orders to establish the required happens-before relation, so all operations on `timestamp` (including the load in `get`) can be relaxed.

Reply:
@mpoeter : oh yes, makes a lot of sense. Thanks for pointing me at it!